src/Core/Cart/CartRuleCalculator.php

```diff
@@ -153,7 +153,7 @@ protected function applyCartRule(CartRuleData $cartRuleData, $withFreeShipping =
                 $cartRowCheapest = null;
                 foreach ($this->cartRows as $cartRow) {
                     $product = $cartRow->getRowData();
-                    if (((($cartRule->reduction_exclude_special && !$product['reduction_applies'])
+                    if (((($cartRule->reduction_exclude_special && isset($product['reduction_applies']) && !$product['reduction_applies'])
```

Comment:
Only a suggestion, not required
```suggestion
                    if (((($cartRule->reduction_exclude_special && !($product['reduction_applies'] ?? true)
```

Reply:
I like it ;)